File: 3.x.x/crates-utopia/http/src/router.rs
use crate::error::{HttpError, Result};
use crate::route::Route;
use parking_lot::RwLock;
use std::collections::HashMap;
use std::fmt;
use std::sync::Arc;

const PLACEHOLDER: &str = ":*:";
const WILDCARD: &str = "*";

#[derive(Clone, Debug)]
pub struct RouteMatch {
    pub route: Arc<Route>,
    pub params: HashMap<String, String>,
}

#[derive(Default)]
struct RouterInner {
    routes: HashMap<String, HashMap<String, Arc<Route>>>,
    param_indexes: Vec<usize>,
    wildcard: Option<Arc<Route>>,
    allow_override: bool,
}

/// Utopia-compatible router (`:param`, `*`, aliases, multi-method).
#[derive(Clone, Default)]
pub struct Router {
    inner: Arc<RwLock<RouterInner>>,
}

impl fmt::Debug for Router {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Router").finish_non_exhaustive()
    }
}

impl Router {
    pub fn new() -> Self {
        let mut routes = HashMap::new();
        for m in ["GET", "POST", "PUT", "PATCH", "DELETE"] {
            routes.insert(m.to_string(), HashMap::new());
        }
        Self {
            inner: Arc::new(RwLock::new(RouterInner {
                routes,
                param_indexes: Vec::new(),
                wildcard: None,
                allow_override: false,
            })),
        }
    }

    pub fn set_allow_override(&self, value: bool) {
        self.inner.write().allow_override = value;
    }

    pub fn get_allow_override(&self) -> bool {
        self.inner.read().allow_override
    }

    pub fn set_wildcard(&self, route: Arc<Route>) {
        self.inner.write().wildcard = Some(route);
    }

    pub fn add_route(&self, route: Arc<Route>) -> Result<()> {
        let (path, params) = prepare_path(route.path());
        let methods = route.methods().to_vec();
        if methods.is_empty() {
            return Err(HttpError::EmptyMethods);
        }
        let mut inner = self.inner.write();
        for method in &methods {
            if !inner.routes.contains_key(method) {
                return Err(HttpError::UnsupportedMethod(method.clone()));
            }
            if inner.routes[method].contains_key(&path) && !inner.allow_override {
                return Err(HttpError::DuplicateRoute {
                    method: method.clone(),
                    path: path.clone(),
                });
            }
        }
        for (key, index) in &params {
            route.set_path_param(key, *index, &path);
            if !inner.param_indexes.contains(index) {
                inner.param_indexes.push(*index);
            }
        }
        for method in methods {
            inner
                .routes
                .get_mut(&method)
                .unwrap()
                .insert(path.clone(), route.clone());
        }
        Ok(())
    }

    pub fn add_route_alias(&self, alias_path: &str, route: Arc<Route>) -> Result<()> {
        let (alias, params) = prepare_path(alias_path);
        let methods = route.methods().to_vec();
        let mut inner = self.inner.write();
        for method in &methods {
            if !inner.routes.contains_key(method) {
                return Err(HttpError::UnsupportedMethod(method.clone()));
            }
            if inner.routes[method].contains_key(&alias) && !inner.allow_override {
                return Err(HttpError::DuplicateRoute {
                    method: method.clone(),
                    path: alias.clone(),
                });
            }
        }
        for (key, index) in &params {
            route.set_path_param(key, *index, &alias);
            if !inner.param_indexes.contains(index) {
                inner.param_indexes.push(*index);
            }
        }
        for method in methods {
            inner
                .routes
                .get_mut(&method)
                .unwrap()
                .insert(alias.clone(), route.clone());
        }
        route.add_alias_path(alias_path);
        Ok(())
    }

    pub fn match_route(&self, method: &str, path: &str) -> Option<RouteMatch> {
        let inner = self.inner.read();
        let method = if method == "HEAD" { "GET" } else { method };
        let Some(table) = inner.routes.get(method) else {
            return inner.wildcard.clone().map(|route| RouteMatch {
                route,
                params: HashMap::new(),
            });
        };

        let parts: Vec<&str> = path.split('/').filter(|s| !s.is_empty()).collect();
        let static_key = parts.join("/");

        // Fast path: exact static template (no placeholder substitution).
        if let Some(route) = table.get(&static_key) {
            let params = route.resolve_params_from_parts(&parts, &static_key);
            return Some(RouteMatch {
                route: route.clone(),
                params,
            });
        }

        let length = parts.len().saturating_sub(1);
        let filtered: Vec<usize> = inner
            .param_indexes
            .iter()
            .copied()
            .filter(|i| *i <= length)
            .collect();

        // Parametric routes only when placeholder indexes exist for this path length.
        if !filtered.is_empty() {
            // Skip the empty sample - already tried as static_key above.
            for sample in combinations(&filtered).into_iter().skip(1) {
                let sample: Vec<usize> = sample.into_iter().filter(|i| *i <= length).collect();
                if sample.is_empty() {
                    continue;
                }
                let mut template_parts = parts.clone();
                for i in &sample {
                    if *i < template_parts.len() {
                        template_parts[*i] = PLACEHOLDER;
                    }
                }
                let template = template_parts.join("/");
                if let Some(route) = table.get(&template) {
                    let params = route.resolve_params_from_parts(&parts, &template);
                    return Some(RouteMatch {
                        route: route.clone(),
                        params,
                    });
                }
            }
        }

        if let Some(route) = table.get(WILDCARD) {
            return Some(RouteMatch {
                route: route.clone(),
                params: HashMap::new(),
            });
        }

        let mut current = String::new();
        for part in &parts {
            current.push_str(part);
            current.push('/');
            let template = format!("{current}{WILDCARD}");
            if let Some(route) = table.get(&template) {
                return Some(RouteMatch {
                    route: route.clone(),
                    params: HashMap::new(),
                });
            }
        }

        inner.wildcard.clone().map(|route| RouteMatch {
            route,
            params: HashMap::new(),
        })
    }

    pub fn reset(&self) {
        let mut inner = self.inner.write();
        *inner = RouterInner {
            routes: {
                let mut routes = HashMap::new();
                for m in ["GET", "POST", "PUT", "PATCH", "DELETE"] {
                    routes.insert(m.to_string(), HashMap::new());
                }
                routes
            },
            param_indexes: Vec::new(),
            wildcard: None,
            allow_override: false,
        };
    }
}

pub fn prepare_path(path: &str) -> (String, HashMap<String, usize>) {
    let parts: Vec<&str> = path.split('/').filter(|s| !s.is_empty()).collect();
    let mut prepare = String::new();
    let mut params = HashMap::new();
    for (key, part) in parts.iter().enumerate() {
        if key != 0 {
            prepare.push('/');
        }
        if let Some(name) = part.strip_prefix(':') {
            prepare.push_str(PLACEHOLDER);
            params.insert(name.to_string(), key);
        } else {
            prepare.push_str(part);
        }
    }
    (prepare, params)
}

fn combinations(set: &[usize]) -> Vec<Vec<usize>> {
    let mut results = vec![vec![]];
    let mut out = vec![vec![]];
    for &element in set {
        let mut next = Vec::new();
        for combination in &results {
            let mut ret = vec![element];
            ret.extend(combination.iter().copied());
            next.push(ret.clone());
            out.push(ret);
        }
        results.extend(next);
    }
    out
}

File: packages/http/src/Http/Router.php
<?php

namespace Utopia\Http;

use Exception;

class Router
{
    /**
     * Placeholder token for params in paths.
     */
    public const string PLACEHOLDER_TOKEN = ':::';
    public const string WILDCARD_TOKEN = '*';

    protected static bool $allowOverride = false;

    /**
     * Method-agnostic wildcard route used when no method-specific route matches.
     */
    protected static ?Route $wildcard = null;

    /**
     * @var array<string,Route[]>
     */
    protected static array $routes = [
        Http::REQUEST_METHOD_GET => [],
        Http::REQUEST_METHOD_POST => [],
        Http::REQUEST_METHOD_PUT => [],
        Http::REQUEST_METHOD_PATCH => [],
        Http::REQUEST_METHOD_DELETE => [],
    ];

    /**
     * Contains the positions of all params in the paths of all registered Routes.
     *
     * @var array<int>
     */
    protected static array $params = [];

    /**
     * Get all registered routes.
     *
     * @return array<string, Route[]>
     */
    public static function getRoutes(): array
    {
        return self::$routes;
    }

    /**
     * Get allow override
     *
     */
    public static function getAllowOverride(): bool
    {
        return self::$allowOverride;
    }

    /**
     * Set Allow override
     *
     */
    public static function setAllowOverride(bool $value): void
    {
        self::$allowOverride = $value;
    }


    /**
     * Add route to router.
     *
     * @throws \Exception
     */
    public static function addRoute(Route $route): void
    {
        [$path, $params] = self::preparePath($route->getPath());
        $methods = $route->getMethods();
        $method = $methods[0] ?? '';
        $additionalMethods = \array_slice($methods, 1);

        if (!\array_key_exists($method, self::$routes)) {
            throw new Exception("Method ({$method}) not supported.");
        }

        if (\array_key_exists($path, self::$routes[$method]) && !self::$allowOverride) {
            throw new Exception("Route for ({$method}:{$path}) already registered.");
        }

        foreach ($additionalMethods as $additionalMethod) {
            if (!\array_key_exists($additionalMethod, self::$routes)) {
                throw new Exception("Method ({$additionalMethod}) not supported.");
            }

            if ($route->getPath() === '') {
                throw new Exception('Additional route methods are not supported for the wildcard route.');
            }

            if (\array_key_exists($path, self::$routes[$additionalMethod]) && !self::$allowOverride) {
                throw new Exception("Route for ({$additionalMethod}:{$path}) already registered.");
            }
        }

        foreach ($params as $key => $index) {
            $route->setPathParam($key, $index, $path);
        }

        self::$routes[$method][$path] = $route;

        foreach ($additionalMethods as $additionalMethod) {
            self::$routes[$additionalMethod][$path] = $route;
        }
    }

    /**
     * Add route to router.
     *
     * @throws \Exception
     */
    public static function addRouteAlias(string $path, Route $route): void
    {
        $methods = $route->getMethods();
        [$alias, $params] = self::preparePath($path);

        foreach ($methods as $method) {
            if (!\array_key_exists($method, self::$routes)) {
                throw new Exception("Method ({$method}) not supported.");
            }

            if (\array_key_exists($alias, self::$routes[$method]) && !self::$allowOverride) {
                throw new Exception("Route for ({$method}:{$alias}) already registered.");
            }
        }

        foreach ($params as $key => $index) {
            $route->setPathParam($key, $index, $alias);
        }

        foreach ($methods as $method) {
            self::$routes[$method][$alias] = $route;
        }
    }

    /**
     * Register a method-agnostic catch-all route, used when nothing else matches.
     */
    public static function setWildcard(?Route $route): void
    {
        self::$wildcard = $route;
    }

    /**
     * Find the route registered for a request's method and path.
     */
    public static function match(string $method, string $path): ?RouteMatch
    {
        if (!\array_key_exists($method, self::$routes)) {
            return self::$wildcard !== null ? new RouteMatch(self::$wildcard, []) : null;
        }

        $parts = array_values(array_filter(explode('/', $path), fn($segment) => $segment !== ''));
        $length = \count($parts) - 1;
        $filteredParams = array_filter(self::$params, fn($i) => $i <= $length);

        foreach (self::combinations($filteredParams) as $sample) {
            $sample = array_filter($sample, fn(int $i) => $i <= $length);
            $template = implode(
                '/',
                array_replace(
                    $parts,
                    array_fill_keys($sample, self::PLACEHOLDER_TOKEN),
                ),
            );

            if (\array_key_exists($template, self::$routes[$method])) {
                $route = self::$routes[$method][$template];
                return new RouteMatch($route, $route->resolveParams($path, $template));
            }
        }

        /**
         * Match root wildcard.
         */
        $template = self::WILDCARD_TOKEN;
        if (\array_key_exists($template, self::$routes[$method])) {
            return new RouteMatch(self::$routes[$method][$template], []);
        }

        /**
         * Match wildcard for path segments.
         */
        foreach ($parts as $part) {
            $current = ($current ?? '') . "{$part}/";
            $template = $current . self::WILDCARD_TOKEN;
            if (\array_key_exists($template, self::$routes[$method])) {
                return new RouteMatch(self::$routes[$method][$template], []);
            }
        }

        if (self::$wildcard !== null) {
            return new RouteMatch(self::$wildcard, []);
        }

        return null;
    }

    /**
     * Get all combinations of the given set.
     *
     * @param array<int, mixed> $set
     * @return iterable<array<int, mixed>>
     */
    protected static function combinations(array $set): iterable
    {
        yield [];

        $results = [[]];

        foreach ($set as $element) {
            foreach ($results as $combination) {
                $ret = array_merge([$element], $combination);
                $results[] = $ret;

                yield $ret;
            }
        }
    }

    /**
     * Prepare path for matching
     *
     * @return array{0: string, 1: array<string, int>}
     */
    public static function preparePath(string $path): array
    {
        $parts = array_values(array_filter(explode('/', $path)));
        $prepare = '';
        $params = [];

        foreach ($parts as $key => $part) {
            if ($key !== 0) {
                $prepare .= '/';
            }

            if (str_starts_with($part, ':')) {
                $prepare .= self::PLACEHOLDER_TOKEN;
                $params[ltrim($part, ':')] = $key;
                if (!\in_array($key, self::$params)) {
                    self::$params[] = $key;
                }
            } else {
                $prepare .= $part;
            }
        }

        return [$prepare, $params];
    }

    /**
     * Reset router
     */
    public static function reset(): void
    {
        self::$params = [];
        self::$wildcard = null;
        self::$allowOverride = false;
        self::$routes = [
            Http::REQUEST_METHOD_GET => [],
            Http::REQUEST_METHOD_POST => [],
            Http::REQUEST_METHOD_PUT => [],
            Http::REQUEST_METHOD_PATCH => [],
            Http::REQUEST_METHOD_DELETE => [],
        ];
    }
}
